config/configmapprovider/provider.go

```diff
@@ -20,42 +20,77 @@ import (
+	//
+	// Should never be called concurrently with itself or with Retrieve.
+	// If ctx is cancelled should return immediately with an error.
+	Shutdown(ctx context.Context) error
 }
 
-// Retrieved holds the result of a call to the Retrieve method of a Session object.
+// Retrieved holds the result of a call to the Retrieve method of a Provider object.
+//
+// The typical usage is the following:
+//
+//		r := mapProvider.Retrieve()
+//		r.Get()
+//		// wait for onChange() to be called.
+//		r.Close()
+//		r.Get()
+//		// wait for onChange() to be called.
+//		r.Close()
+//		// repeat Get/wait/Close cycle until it is time to shut down the Collector process.
+//		// ...
+//		mapProvider.Shutdown()
 type Retrieved interface {
-	// Get returns the Map.
-	Get() *config.Map
-}
+	// Get returns the config Map.
+	// If Close is called before Get or concurrently with Get then Get
+	// should return immediately with ErrSessionClosed error.
+	// Should never be called concurrently with itself.
+	// If ctx is cancelled should return immediately with an error.
+	Get(ctx context.Context) (*config.Map, error)
 
-// WatchableRetrieved is an extension for Retrieved that if implemented,
-// the Retrieved value supports monitoring for updates.
-type WatchableRetrieved interface {
-	Retrieved
-
-	// WatchForUpdate waits for updates on any of the values retrieved from config sources.
-	// It blocks until configuration updates are received and can
-	// return an error if anything fails. WatchForUpdate is used once during the
-	// first evaluation of the configuration and is not used to watch configuration
-	// changes continuously.
-	WatchForUpdate() error
-
-	// Close signals that the configuration for which it was used to retrieve values is no longer in use
-	// and the object should close and release any watchers that it may have created.
+	// Close signals that the configuration for which it was used to retrieve values is
+	// no longer in use and the object should close and release any watchers that it
+	// may have created.
+	//
 	// This method must be called when the service ends, either in case of success or error.
-	// The method may be called while WatchForUpdate() class is in progress and is blocked.
-	// In that case WatchForUpdate() method must abort as soon as possible and return ErrSessionClosed.
+	//
+	// Should never be called concurrently with itself.
+	// May be called before, after or concurrently with Get.
+	// If ctx is cancelled should return immediately with an error.
+	//
+	// Calling Close on an already closed object should have no effect and should return nil.
 	Close(ctx context.Context) error
 }
+
+// ChangeEvent describes the particular change event that happened with the config.
+// TODO: see if this can be eliminated.
+type ChangeEvent struct {
+	// Error is nil if the config is changed and needs to be re-fetched using Get.
+	// It is set to configsource.ErrSessionClosed if Close was called.
```

Comment:
Not sure we need this. When Close is called we can simply just not call the `onChange` callback. We just need to ensure that we will not call the callback after Close returns.

Reply:
Yes, see my TODO comment above. Going to look into that separately.

Reply:
We still need an "error" for the case when the watcher fails something like an "errorHandler" for the watcher. We either have to funcs "onChange" and "onError" or we have this and if `err != nil` probably we should close the service? Because I don't know how to recover similar with the Host.ReportError()?

So I would keep for the moment the error.